merkle-tree/concurrent/src/lib.rs

```diff
@@ -0,0 +1,360 @@
+
+    /// Increments the changelog counter. If it reaches the limit, it starts
+    /// from the beginning.
+    fn inc_current_changelog_index(&mut self) {
+        // NOTE(vadorovsky): Apparenty, Rust doesn't have `checked_remainder`
+        // or anything like that.
+        self.current_changelog_index = if MAX_CHANGELOG > 0 {
+            (self.current_changelog_index + 1) % MAX_CHANGELOG as u64
+        } else {
+            0
+        };
+    }
+
+    /// Increments the root counter. If it reaches the limit, it starts from
+    /// the beginning.
+    fn inc_current_root_index(&mut self) {
+        self.current_root_index = (self.current_root_index + 1) % MAX_ROOTS as u64;
+    }
+
+    /// Returns the index of the current changelog entry.
+    pub fn changelog_index(&self) -> usize {
+        self.current_changelog_index as usize
+    }
+
+    /// Returns the index of the current root in the tree's root buffer.
+    pub fn root_index(&self) -> usize {
+        self.current_root_index as usize
+    }
+
+    /// Returns the current root.
+    pub fn root(&self) -> Result<[u8; 32], HasherError> {
+        self.roots
+            .get(self.current_root_index as usize)
+            .ok_or(HasherError::RootHigherThanMax)
+            .map(|&value| value)
+    }
+
+    /// Returns an intersection index in the changelog entry which affects the
+    ///
+    /// Determining it can be done by taking a XOR of the leaf index (which was
+    /// directly updated in the changelog entry) and the leaf index we are
+    /// trying to update.
+    ///
+    /// The number of bytes in the binary representations of the indexes is
+    /// determined by the height of the tree. For example, for the tree with
+    /// height 4, update attempt of leaf under index 2 and changelog affecting
+    /// index 4, critbit would be:
+    ///
+    /// 2 ^ 4 = 0b_0010 ^ 0b_0100 = 0b_0110 = 6
+    fn intersection_index(&self, leaf_index: usize, changelog_entry_index: usize) -> usize {
+        let padding = 64 - MAX_HEIGHT;
+        let common_path_len =
+            ((leaf_index ^ changelog_entry_index) << padding).leading_zeros() as usize;
+
+        (MAX_HEIGHT - 1) - common_path_len
+    }
+
+    /// Returns an updated Merkle proof and an updated leaf after potential
+    /// changes which might have been made after the change referenced by the
+    /// proof.
+    ///
+    /// It does so by checking whether there are any new changelog entries and
+    /// whether they contain changes which affect the current proof. To be
+    /// precise, for each changelog entry, it's done in the following steps:
+    ///
+    /// * Check if the changelog entry was directly updating the `leaf_index`
+    ///   we are trying to update.
+    ///   * If no (we check that condition first, since it's more likely),
+    ///     it means that there is a change affecting the proof, but not the
+    ///     leaf.
+    ///     Check which leaf from our proof was affected by the change (using
```

Comment:
```suggestion
    ///     Check which element of our proof was affected by the change (using
```

Reply:
It's not a leaf, but an element of the Merkle proof. An alternative name might be node, although I like element better.